share/eds/blockstore.go

```diff
@@ -0,0 +1,195 @@
+// clarification, an implementation of Put does not make sense in this context because it is unclear which CAR file the
+// block should be written to.
+func (bs *Blockstore) Put(ctx context.Context, block blocks.Block) error {
+	return nil
+}
+
+// PutMany is a noop on the EDS blockstore, but it does not return an error because it is called by bitswap. For
+// clarification, an implementation of PutMany does not make sense in this context because it is unclear which CAR file
+// the blocks should be written to.
+func (bs *Blockstore) PutMany(ctx context.Context, blocks []blocks.Block) error {
+	return nil
+}
+
+// AllKeysChan is a noop on the EDS blockstore because the keys are not stored in a single CAR file.
+func (bs *Blockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	return nil, ErrUnsupportedOperation
+}
+
+// HashOnRead is a noop on the EDS blockstore but an error cannot be returned due to the method signature from the
+// blockstore interface.
+func (bs *Blockstore) HashOnRead(bool) {
+	log.Warnf("HashOnRead is a noop on the EDS blockstore")
+}
+
+func (bs *Blockstore) getReadOnlyBlockstore(ctx context.Context, cid cid.Cid) (dagstore.ReadBlockstore, error) {
+	keys, err := bs.store.dgstr.ShardsContainingMultihash(ctx, cid.Hash())
+	if err != nil {
+		return nil, err
+	}
+	if len(keys) != 1 {
+		return nil, ErrInvalidShardCount
+	}
+
+	// try to fetch from cache
+	shardKey := keys[0]
+	blockstr, err := bs.readFromBSCache(shardKey)
+	if err == nil && blockstr != nil {
+		return blockstr, nil
+	}
+
+	// wasn't found in cache, so acquire it and add to cache
+	ch := make(chan dagstore.ShardResult, 1)
+	err = bs.store.dgstr.AcquireShard(ctx, shardKey, ch, dagstore.AcquireOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case res := <-ch:
+		if res.Error != nil {
+			return nil, res.Error
+		}
+		return bs.addToBSCache(shardKey, res.Accessor)
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (bs *Blockstore) readFromBSCache(shardContainingCid shard.Key) (dagstore.ReadBlockstore, error) {
+	lk := &bs.bsStripedLocks[shardKeyToStriped(shardContainingCid)]
+	lk.Lock()
+	defer lk.Unlock()
+
+	// We've already ensured that the given shard has the cid/multihash we are looking for.
+	val, ok := bs.blockstoreCache.Get(shardContainingCid)
+	if !ok {
+		return nil, errors.New("not found in cache")
+	}
+
+	rbs, ok := val.(*accessorWithBlockstore)
+	if !ok {
+		return nil, errors.New("couldn't cast value from cache to accessor with blockstore")
+	}
+	return rbs.bs, nil
+}
+
+func (bs *Blockstore) addToBSCache(
+	shardContainingCid shard.Key,
+	accessor *dagstore.ShardAccessor,
+) (dagstore.ReadBlockstore, error) {
+	lk := &bs.bsStripedLocks[shardKeyToStriped(shardContainingCid)]
+	lk.Lock()
+	defer lk.Unlock()
+
+	blockStore, err := accessor.Blockstore()
+	if err != nil {
+		return nil, err
+	}
+
+	bs.blockstoreCache.Add(shardContainingCid, &accessorWithBlockstore{
+		bs: blockStore,
+		sa: accessor,
+	})
```

Comment:
Small optimisation here could be to obtain lock after `blockStore` is returned, if lock doesn't need to protect `accessor`.